File: app/modules/planning/domain/calculations/patterns/rebalance.py
"""Rebalance pattern generator.

Sell overweight positions and buy underweight areas.
"""

from typing import Any, Dict, List

from app.modules.planning.domain.calculations.patterns.base import (
    PatternGenerator,
    pattern_generator_registry,
)
from app.modules.planning.domain.models import ActionCandidate


class RebalancePattern(PatternGenerator):
    """Rebalance: Sell overweight + buy underweight."""

    @property
    def name(self) -> str:
        return "rebalance"

    def default_params(self) -> Dict[str, Any]:
        return {
            "max_depth": 5,
            "available_cash_eur": 0.0,
        }

    def generate(
        self,
        opportunities: Dict[str, List[ActionCandidate]],
        params: Dict[str, Any],
    ) -> List[List[ActionCandidate]]:
        """
        Generate rebalance pattern: Sell overweight, buy underweight.

        Args:
            opportunities: Dict of categorized opportunities
            params: Pattern parameters

        Returns:
            List containing one rebalancing sequence, or empty list
        """
        rebalance_sells = opportunities.get("rebalance_sells", [])
        if not rebalance_sells:
            return []

        max_depth = params.get("max_depth", 5)
        available_cash = params.get("available_cash_eur", 0.0)

        # Take top rebalance sells
        sequence: List[ActionCandidate] = rebalance_sells[
            : min(len(rebalance_sells), max_depth)
        ]

        # Calculate cash from sells
        cash_from_sells = sum(c.value_eur for c in sequence)
        total_cash = available_cash + cash_from_sells

        # Buy underweight areas
        rebalance_buys = opportunities.get("rebalance_buys", [])
        for candidate in rebalance_buys:
            if candidate.value_eur <= total_cash and len(sequence) < max_depth:
                sequence.append(candidate)
                total_cash -= candidate.value_eur

        return [sequence] if sequence else []


# Auto-register this pattern
_rebalance_pattern = RebalancePattern()
pattern_generator_registry.register(_rebalance_pattern.name, _rebalance_pattern)

File: trader/internal/modules/sequences/patterns/rebalance.go
package patterns

import (
	"github.com/aristath/arduino-trader/internal/modules/planning/domain"
	"github.com/rs/zerolog"
)

// RebalancePattern generates sequences that rebalance the portfolio by selling overweight
// and buying underweight positions.
type RebalancePattern struct {
	*BasePattern
}

// NewRebalancePattern creates a new rebalance pattern generator.
func NewRebalancePattern(log zerolog.Logger) *RebalancePattern {
	return &RebalancePattern{
		BasePattern: NewBasePattern(log, "rebalance"),
	}
}

// Name returns the pattern name.
func (p *RebalancePattern) Name() string {
	return "rebalance"
}

// Generate creates rebalancing sequences.
func (p *RebalancePattern) Generate(
	opportunities domain.OpportunitiesByCategory,
	params map[string]interface{},
) ([]domain.ActionSequence, error) {
	// Parameters
	maxSequences := GetIntParam(params, "max_sequences", 5)
	sellFirst := GetBoolParam(params, "sell_first", true) // Sell before buy for cash generation

	p.log.Debug().
		Int("max_sequences", maxSequences).
		Bool("sell_first", sellFirst).
		Msg("Generating rebalance sequences")

	var sequences []domain.ActionSequence

	// Get rebalance opportunities
	rebalanceSells, hasSells := opportunities[domain.OpportunityCategoryRebalanceSells]
	rebalanceBuys, hasBuys := opportunities[domain.OpportunityCategoryRebalanceBuys]

	if !hasSells && !hasBuys {
		p.log.Debug().Msg("No rebalance opportunities found")
		return sequences, nil
	}

	p.log.Debug().
		Int("sells", len(rebalanceSells)).
		Int("buys", len(rebalanceBuys)).
		Msg("Rebalance opportunities")

	// Create sell-only sequences
	for i, sellCandidate := range rebalanceSells {
		if maxSequences > 0 && i >= maxSequences {
			break
		}
		sequence := CreateSequence([]domain.ActionCandidate{sellCandidate}, "rebalance")
		sequences = append(sequences, sequence)
	}

	// Create buy-only sequences
	for i, buyCandidate := range rebalanceBuys {
		if maxSequences > 0 && i >= maxSequences {
			break
		}
		sequence := CreateSequence([]domain.ActionCandidate{buyCandidate}, "rebalance")
		sequences = append(sequences, sequence)
	}

	// Create combined sell-then-buy sequences
	if hasSells && hasBuys {
		count := 0
		for _, sellCandidate := range rebalanceSells {
			for _, buyCandidate := range rebalanceBuys {
				if maxSequences > 0 && count >= maxSequences {
					break
				}

				var actions []domain.ActionCandidate
				if sellFirst {
					actions = []domain.ActionCandidate{sellCandidate, buyCandidate}
				} else {
					actions = []domain.ActionCandidate{buyCandidate, sellCandidate}
				}

				sequence := CreateSequence(actions, "rebalance")
				sequences = append(sequences, sequence)
				count++
			}
			if maxSequences > 0 && count >= maxSequences {
				break
			}
		}
	}

	p.log.Info().
		Int("sequences", len(sequences)).
		Msg("Rebalance sequences generated")

	return sequences, nil
}
